### PR TITLE
Add CachingOptimizer in doc

### PR DESCRIPTION
**Pull request by frapac**
Solve #770 

Note that:
- this is a first attempt. If I miss anything, I welcome any feedback.
- I am not sure of the position of the `CachingOptimizer` paragraph in the page `apireference.md`: shall `CachingOptimizer` precede `Allocate-Load API`?

### CODE REVIEW

File: docs/src/apireference.md [hunk 1]
```diff
@@ -547,6 +547,43 @@ Utilities.load
...
 ```
 
+### Caching optimizer
+
+Some solvers do not support incremental definition of optimization
+models. Nevertheless, you are still able to build incrementally an optimization
+model with such solvers. MathOptInterface allows to use a utility,
+[`Utilities.CachingOptimizer`](@ref), that will store in a [`ModelLike`](@ref) `model_cache`
+the optimization model during its incremental definition. Once the
+model completely defined, the `CachingOptimizer` specifies all problem's
```
Comment (blegat) [inline docs/src/apireference.md]: `model` -> `model is`

File: docs/src/apireference.md [hunk 2]
```diff
@@ -547,6 +547,43 @@ Utilities.load
...
+information to the underlying solver, all in once.
+
+The way to operate a `CachingOptimizer` is as follow.
+1) Suppose you want to cache a (empty) [`AbstractOptimizer`](@ref) `optimizer`.
```
Comment (blegat) [inline docs/src/apireference.md]: `a` -> `an`

File: docs/src/apireference.md [hunk 3]
```diff
@@ -547,6 +547,43 @@ Utilities.load
...
+2) Define a [`ModelLike`](@ref) `model_cache` that will store the incremental definition of the model.
+3) Cache `optimizer` with `model_cache` by calling `CachingOptimizer(model_cache, optimizer)`. This method returns an empty `AbstractOptimizer` `cached_optimizer`.
+4) Define your optimization model as usual by using `cached_optimizer` instead of `optimizer`. This will allocate the model directly to `model_cache` instead of `optimizer`. For instance, to add two variables in the model:
+```julia
+MOI.add_variables(cached_optimizer, 2)
+```
+5) Once the model properly defined, call [`Utilities.attach_optimizer`](@ref) to copy the optimization model all in once inside `optimizer`. This method uses the MathOptInterface method [`copy_to`](@ref) to copy information from `model_cache` to `optimizer`.
```
Comment (blegat) [inline docs/src/apireference.md]: `model` -> `model is`

File: docs/src/apireference.md [hunk 4]
```diff
@@ -547,6 +547,43 @@ Utilities.load
...
+
+Note that:
+- [`Utilities.drop_optimizer`](@ref) drops the underlying `optimizer` from `cached_optimizer`, without emptying it.
+- [`Utilities.reset_optimizer`](@ref) empties `optimizer` inside `cached_optimizer`, without droping it.
+
+Bridging correctly a solver with a `CachingOptimizer` requires to
+implement properly the [Allocate-Load API](@ref) inside the solver's
```
Comment (blegat) [inline docs/src/apireference.md]: Not if you add the bridges on top of the CachingOptimizer. If the solver does not implement incremental copy or allocate-load, then it needs a cache between the bridges and the optimizer. You can make a reference to http://www.juliaopt.org/MathOptInterface.jl/dev/apimanual/#Implementing-copy-1

File: docs/src/apireference.md (hunk 3 above)
Comment (blegat) [inline docs/src/apireference.md]: `attach_optmizer` is called automatically if the mode is `AUTOMATIC`

File: docs/src/apireference.md [hunk 5]
```diff
@@ -547,6 +547,43 @@ Utilities.load
...
+model with such solvers. MathOptInterface allows to use a utility,
+[`Utilities.CachingOptimizer`](@ref), that will store in a [`ModelLike`](@ref) `model_cache`
+the optimization model during its incremental definition. Once the
+model completely defined, the `CachingOptimizer` specifies all problem's
+information to the underlying solver, all in once.
+
+The way to operate a `CachingOptimizer` is as follow.
+1) Suppose you want to cache a (empty) [`AbstractOptimizer`](@ref) `optimizer`.
+2) Define a [`ModelLike`](@ref) `model_cache` that will store the incremental definition of the model.
+3) Cache `optimizer` with `model_cache` by calling `CachingOptimizer(model_cache, optimizer)`. This method returns an empty `AbstractOptimizer` `cached_optimizer`.
```
Comment (blegat) [inline docs/src/apireference.md]: There is also a constructor where you only give the cache and the mode and then to give the optimizer with `attach_optimizer`.

File: docs/src/apireference.md [hunk 6]
```diff
@@ -554,26 +554,44 @@ models. Nevertheless, you are still able to build incrementally an optimization
...
+* [`Utilities.reset_optimizer`](@ref) empties `optimizer` inside
+  `cached_optimizer`, without droping it. The state of `cached_optimizer`
+  is set to `EMPTY_OPTIMIZER` after the call.
+
+The way to operate a `CachingOptimizer` depends whether the mode
+is set to `AUTOMATIC` or to `MANUAL`.
+* In `MANUAL` mode, the state of the `CachingOptimizer` changes only
+  if the methods [`Utilities.attach_optimizer`](@ref),
+  [`Utilities.reset_optimizer`](@ref) or [`Utilities.drop_optimizer`](@ref)
+  are being called.  Any unattended operation results in an error.
```
Comment (blegat) [inline docs/src/apireference.md]: Two spaces after dot

File: docs/src/apireference.md [hunk 7]
```diff
@@ -554,26 +554,44 @@ models. Nevertheless, you are still able to build incrementally an optimization
...
+* In `AUTOMATIC` mode, the state of the `CachingOptimizer` changes when
+  necessary. Any modification not supported by the solver (e.g. dropping
+  a constraint) results in a drop to the state `EMPTY_OPTIMIZER`.
+
+Wrapping correctly a solver with a `CachingOptimizer` requires to
+implement properly the [Allocate-Load API](@ref) inside the
```
Comment (blegat) [inline docs/src/apireference.md]: The solver should support `MOI.copy_to` but `MOI.copy_to` could
* either use the `add_constraint`, `add_variable`, ... API hence it uses `default_copy_to`
* either use `allocate_...` `load...` hence it used `allocate_load`
* or simply implement a custom `MOI.copy_to`.

This is not true that the solver requires to implement Allocate-Load,

File: docs/src/apireference.md [hunk 8]
```diff
@@ -554,26 +554,44 @@ models. Nevertheless, you are still able to build incrementally an optimization
...
+MathOptInterface wrapper of the solver.
+Note that bridges do not yet implement the Allocate-Load API. A `CachingOptimizer`
```
Comment (blegat) [inline docs/src/apireference.md]: Use a new paragraph

File: docs/src/apireference.md [hunk 9]
```diff
@@ -554,26 +554,44 @@ models. Nevertheless, you are still able to build incrementally an optimization
...
+is required if you are using a bridged solver and want to use the
```
Comment (blegat) [inline docs/src/apireference.md]: solver -> optimizer

File: docs/src/apireference.md [hunk 10]
```diff
@@ -554,26 +554,44 @@ models. Nevertheless, you are still able to build incrementally an optimization
...
+Allocate-Load API, as explained in the section [Implementing copy](@ref).
```
Comment (blegat) [inline docs/src/apireference.md]: This is not if the user want to use the allocate-load API, it's simply that bridge optimizers requires the underlying model to support `add_constraint`, `add_variable`, ...
So if the underlying model does not support it then it needs a cache.
In the future, when the bridge optimizers implement Allocate-Load, the underlying layer could either implement `add_...` or `allocate_...`/`load_...` but if it only implements a custom `copy_to`, it needs a cache

File: docs/src/apireference.md (hunk 7 above)
Comment (frapac) [reply docs/src/apireference.md]: You made a point. I understand that this paragraph is unclear. Maybe we could just remove it?

Comment (blegat) [reply docs/src/apireference.md]: Yes, it's tricky to explain and should already be covered in "Implementing copy"

Comment (frapac) [reply docs/src/apireference.md]: I agree :) I will make a reference to "Implementing copy" instead, and rescope this PR.

**Comment by frapac** [comment]
Thanks for your review!

File: docs/src/apireference.md [hunk 11]
```diff
@@ -547,6 +547,58 @@ Utilities.load
 Utilities.load_constraint
 ```
 
+### Caching optimizer
+
+Some solvers do not support incremental definition of optimization
+models. Nevertheless, you are still able to build incrementally an optimization
+model with such solvers. MathOptInterface allows to use a utility,
```
Comment (mlubin) [inline docs/src/apireference.md]: s/allows to use/provides/

File: docs/src/apireference.md [hunk 12]
```diff
@@ -547,6 +547,58 @@ Utilities.load
...
+[`Utilities.CachingOptimizer`](@ref), that will store in a [`ModelLike`](@ref) `model_cache`
```
Comment (mlubin) [inline docs/src/apireference.md]: Is this `model_cache` supposed to be here? The sentence makes sense without it.

File: docs/src/apireference.md [hunk 13]
```diff
@@ -547,6 +547,58 @@ Utilities.load
...
+the optimization model during its incremental definition. Once the
+model is completely defined, the `CachingOptimizer` specifies all problem's
```
Comment (mlubin) [inline docs/src/apireference.md]: s/all problem's information/all problem information/

File: docs/src/apireference.md [hunk 14]
```diff
@@ -547,6 +547,58 @@ Utilities.load
...
+information to the underlying solver, all in once.
```
Comment (mlubin) [inline docs/src/apireference.md]: s/all in once/all at once/

File: docs/src/apireference.md [hunk 15]
```diff
@@ -547,6 +547,58 @@ Utilities.load
...
+
+The function [`Utilities.state`](@ref) allows to query the state
+of the optimizer cached inside a `CachingOptimizer`. The state
+could be:
+* `NO_OPTIMIZER`, if no optimizer is attached;
+* `EMPTY_OPTIMIZER`, if the attached optimizer is empty;
+* `ATTACHED_OPTIMIZER`, if the attached optimizer is synchronized with the
+  cached model defined in `CachingOptimizer`.
+
+Note that there exists methods to modify the attached optimizer:
```
Comment (mlubin) [inline docs/src/apireference.md]: I would word this as:
"The following methods modify the state of the attached optimizer:"

**Comment by frapac** [comment]
Thanks for the comments!

**Pull request merged**